### ISSUE TITLE
(Alpha 2) parse allows invalid day of months 

### ISSUE THREAD

**Issue by bestguy** [date-fns/date-fns#549]
The alpha 2 parse allows day of month to be 0 or greater than 31, and returns dates in the previous or next months.  I'd expect these to be invalid dates, or at least have an option to disable that behavior:

Correct:

```
> parse('12/25/1971', 'MM/DD/YYYY', new Date())
1971-12-25T08:00:00.000Z
> parse('12/2/1971', 'MM/DD/YYYY', new Date())
Invalid Date
```

Incorrect:

```
> parse('12/00/1971', 'MM/DD/YYYY', new Date())
1971-11-30T08:00:00.000Z
> parse('12/42/1971', 'MM/DD/YYYY', new Date())
1972-01-11T08:00:00.000Z
```

**Comment by logaretm** [comment]
Hello, we wanted to use `date-fns` instead of `moment` for various reasons for `vee-validate`, and for validation this is a big problem but I've found a way around it and you may find it useful:

```js
export const parseDate = (date, format) => {
  const parsed = parse(date, format, new Date());

  // if date is not valid or the formatted output after parsing does not match
  // the string value passed in (avoids overflows)
  if (!isValid(parsed) || formatDate(parsed, format) !== date) {
    return null;
  }

  return parsed;
};
```

It is very straightforward but I think it does the job for now, it parses the date given by the user first, checks if its valid, then formats it back to the same format and checks it against the user input. If they don't match then probably the date was overflown or invalid from validation standpoint.

**Comment by glintik** [comment]
Hi guys,

I can confirm an issue with `parse`. It should return "Invalid Date", if date is not in supplied format, my test:

> var parseDate = require('date-fns/parseDate');
> var dateString = '09-15-2017', format = 'YYYY-MM-DD';
> 
> console.log(`Parsing ${dateString} with format ${format}:`, parse(dateString, format, new Date()));

Current output:
`Parsing 09-15-2017 with format YYYY-MM-DD: 0010-03-19T22:00:00.000Z`

Expected:
`
Parsing 09-15-2017 with format YYYY-MM-DD: Invalid Date
`

**Comment by robport** [comment]
+1 - is this likely to get implemented?

**Comment by leshakoss** [comment]
Fixed in [v2.0.0-alpha.14](https://gist.github.com/kossnocorp/a307a464760b405bb78ef5020a4ab136#v200-alpha14) in PR  #815
